Shared/Samples/Sketch on the map/SketchOnTheMapView.swift

```diff
@@ -0,0 +1,289 @@
+            } label: {
+                Label("Cancel Sketch", systemImage: "xmark")
+            }
+        }
+    }
+}
+
+/// An object that acts as a view model for the geometry editor menu.
+@MainActor class GeometryEditorMenuModel: ObservableObject {
+    /// The geometry editor.
+    let geometryEditor: GeometryEditor
+    
+    /// The graphics overlay used to save geometries to.
+    let graphicsOverlay: GraphicsOverlay
+    
+    /// A Boolean value indicating if the geometry editor can perform an undo.
+    @Published private(set) var canUndo = false
+    
+    /// A Boolean value indicating if the geometry editor can perform a redo.
+    @Published private(set) var canRedo = false
+    
+    /// The currently selected element.
+    @Published private(set) var selection: GeometryEditorElement?
+    
+    /// A Boolean value indicating if the geometry can be saved to a graphics overlay.
+    @Published private(set) var canSave = false
+    
+    /// A Boolean value indicating if the geometry can be cleared from the geometry editor.
+    @Published private(set) var canClearCurrentSketch = false
+    
+    /// A Boolean value indicating if the saved sketches can be cleared.
+    @Published private(set) var canClearSavedSketches = false
+    
+    /// The current geometry of the geometry editor.
+    @Published private(set) var geometry: Geometry? {
+        didSet {
+            canUndo = geometryEditor.canUndo
+            canRedo = geometryEditor.canRedo
+            canClearCurrentSketch = geometry.map { !$0.isEmpty } ?? false
+            canSave = sketchIsValidForSaving
+        }
+    }
+    
+    /// A Boolean value indicating if the geometry editor has started.
+    @Published var isStarted = false
+    
+    /// Creates the geometry menu with a geometry editor.
+    /// - Parameter geometryEditor: The geometry editor that the menu should interact with.
+    /// - Parameter graphicsOverlay: The graphics overlay that is used to save geometries to.
+    init(geometryEditor: GeometryEditor, graphicsOverlay: GraphicsOverlay) {
+        self.geometryEditor = geometryEditor
+        self.graphicsOverlay = graphicsOverlay
+        
+        Task { [weak self, geometryEditor] in
+            for await geometry in geometryEditor.$geometry {
+                self?.geometry = geometry
+            }
+        }
+        Task { [weak self, geometryEditor] in
+            for await selection in geometryEditor.$selectedElement {
+                self?.selection = selection
+            }
+        }
+    }
+    
+    /// Saves the current geometry to the graphics overlay and stops editing.
+    /// Precondition: `canSave`
+    func save() {
+        let geometry = geometryEditor.geometry!
+        let graphic = Graphic(geometry: geometry, symbol: symbol(for: geometry))
+        graphicsOverlay.addGraphic(graphic)
+        stop()
+        canClearSavedSketches = true
+    }
+    
+    /// Clears all the saved sketches on the graphics overlay.
```

Comment:
I'm wondering if the `save()` method above has a precondition, this can also have 

`/// - Precondition: canClearSavedSketches`

Reply:
it's not necessary in the same way.  The save method does a force-unwrap of the geometry and that's why the precondition is necessary.